os/compression/compress_read.c

```diff
@@ -379,14 +559,55 @@ int compress_init(int filfd, uint16_t offset, off_t *filelen)
 	/* Assign file length as that of uncompressed file */
 	*filelen = compression_header.binary_size;
 
+	/* Set number of blocks to use for caching */
+	if (CONFIG_DECOMPRESSION_CACHE_BLOCKS < (CUTOFF_RATIO_CACHE_BLOCKS) * (compression_header.sections - 1))
+		number_blocks_caching = CONFIG_DECOMPRESSION_CACHE_BLOCKS;
+	else
+		number_blocks_caching = ((CUTOFF_RATIO_CACHE_BLOCKS) * (compression_header.sections - 1));
+
+	/* Initialize max_accesed_count to 0 */
+	max_accessed_count = 0;
+
 #if CONFIG_COMPRESSION_TYPE == 1
-	/* Allocating memory for read and out buffer to be used for LZMA decompression */
+	/* Allocating memory for compressed_data_buffer and block_caches to be used for LZMA decompression */
 	if (compression_header.compression_format == COMPRESSION_TYPE_LZMA) {
-		buffers.read_buffer = (unsigned char *)malloc(compression_header.blocksize + 5);
-		buffers.out_buffer = (unsigned char *)malloc(compression_header.blocksize);
+
+		compressed_data_buffer = (unsigned char *)kmm_malloc(compression_header.blocksize + 5);
+		if (!compressed_data_buffer) {
+			bmdbg("Failed kmm_malloc for read buffer\n");
+			return -ENOMEM;
+		}
+
+		blockcache = (block_cache_t *)kmm_malloc(number_blocks_caching * sizeof(block_cache_t));
+		if (!blockcache) {
+			bmdbg("Failed kmm_malloc for blockcache\n");
+			return -ENOMEM;
+		}
+
+		for (int i = 0; i < number_blocks_caching; i++) {
+			blockcache[i].out_buffer = (unsigned char *)kmm_malloc(compression_header.blocksize);
+			if (!blockcache[i].out_buffer) {
+				bmdbg("Failed kmm_malloc for blockcache's out_buffer\n");
+				return -ENOMEM;
```

Comment:
Is it okay not to free the `compressed_data_buffer` and `blockcache`?

Reply:
@jeongchanKim, Memory for these will get free in compress_uninit.